A new runner for DRMAA (currently UNIVA)

bernt-matthias opened the pull request:
Reimplementation of the DRMAA runner inspired by the SLURM runner.
Currently tested only for the UNIVA grid engine.

This solves the problem that the current DRMAAJobRunner does
not work when jobs are submitted as real user (because jobs that are
started in a different drmaa session can not be accessed from the
session that is open in galaxy):
- this is done by resorting to command line tools qstat and qacct if
the drmaa library can not be used to check the job status and to get run
time information.
- this has the additional advantage that if the drmaa library
functions are not working (DRMAAJobRunner had implemented a repeated
checking to handle this problem) the runner can still use the command
line tools.

Furthermore (in contrast to the original drmaa runner) the new one
tests for run time and memory violations:
- memory violations are determined by checking the stderr output and
by comparing the used and the requested memory
- run time violations are determined by checking the signal that
killed the job and by comparing the used and the requested run time
Where the used memory and time are determined with drmaa.wait() or
qacct

Open (or better perspective):
- adaptions to other grid engines. the current implementation (the
  command line calls and result parsing) might be specific for the
  Univa grid engine. to include other GEs one could determine the
  GE (+ version) and make the calls and result parsing depending
  on this.

In addition that the job template is loaded in the external runner before
the user is changed. otherwise the cluster_files_directory needs to be
readable by all users.

bernt-matthias committed:
A new runner for DRMAA (currently UNIVA)

Reimplementation of the DRMAA runner inspired by the SLURM runner.
Currently tested only for the UNIVA grid engine.

This solves the problem that the current DRMAAJobRunner does
not work when jobs are submitted as real user (because jobs that are
started in a different drmaa session can not be accessed from the
session that is open in galaxy):
- this is done by resorting to command line tools qstat and qacct if
the drmaa library can not be used to check the job status and to get run
time information.
- this has the additional advantage that if the drmaa library
functions are not working (DRMAAJobRunner had implemented a repeated
checking to handle this problem) the runner can still use the command
line tools.

Furthermore (in contrast to the original drmaa runner) the new one
tests for run time and memory violations:
- memory violations are determined by checking the stderr output and
by comparing the used and the requested memory
- run time violations are determined by checking the signal that
killed the job and by comparing the used and the requested run time
Where the used memory and time are determined with drmaa.wait() or
qacct

Open (or better perspective):
- adaptions to other grid engines. the current implementation (the
  command line calls and result parsing) might be specific for the
  Univa grid engine. to include other GEs one could determine the
  GE (+ version) and make the calls and result parsing depending
  on this.

In addition that the job template is loaded in the external runner before
the user is changed. otherwise the cluster_files_directory needs to be
readable by all users.

bernt-matthias committed:
A new runner for DRMAA (currently UNIVA)

Reimplementation of the DRMAA runner inspired by the SLURM runner.
Currently tested only for the UNIVA grid engine.

This solves the problem that the current DRMAAJobRunner does
not work when jobs are submitted as real user (because jobs that are
started in a different drmaa session can not be accessed from the
session that is open in galaxy):
- this is done by resorting to command line tools qstat and qacct if
the drmaa library can not be used to check the job status and to get run
time information.
- this has the additional advantage that if the drmaa library
functions are not working (DRMAAJobRunner had implemented a repeated
checking to handle this problem) the runner can still use the command
line tools.

Furthermore (in contrast to the original drmaa runner) the new one
tests for run time and memory violations:
- memory violations are determined by checking the stderr output and
by comparing the used and the requested memory
- run time violations are determined by checking the signal that
killed the job and by comparing the used and the requested run time
Where the used memory and time are determined with drmaa.wait() or
qacct

Open (or better perspective):
- adaptions to other grid engines. the current implementation (the
  command line calls and result parsing) might be specific for the
  Univa grid engine. to include other GEs one could determine the
  GE (+ version) and make the calls and result parsing depending
  on this.

In addition that the job template is loaded in the external runner before
the user is changed. otherwise the cluster_files_directory needs to be
readable by all users.

bernt-matthias committed:
A new runner for DRMAA (currently UNIVA)

Reimplementation of the DRMAA runner inspired by the SLURM runner.
Currently tested only for the UNIVA grid engine.

This solves the problem that the current DRMAAJobRunner does
not work when jobs are submitted as real user (because jobs that are
started in a different drmaa session can not be accessed from the
session that is open in galaxy):
- this is done by resorting to command line tools qstat and qacct if
the drmaa library can not be used to check the job status and to get run
time information.
- this has the additional advantage that if the drmaa library
functions are not working (DRMAAJobRunner had implemented a repeated
checking to handle this problem) the runner can still use the command
line tools.

Furthermore (in contrast to the original drmaa runner) the new one
tests for run time and memory violations:
- memory violations are determined by checking the stderr output and
by comparing the used and the requested memory
- run time violations are determined by checking the signal that
killed the job and by comparing the used and the requested run time
Where the used memory and time are determined with drmaa.wait() or
qacct

Open (or better perspective):
- adaptions to other grid engines. the current implementation (the
  command line calls and result parsing) might be specific for the
  Univa grid engine. to include other GEs one could determine the
  GE (+ version) and make the calls and result parsing depending
  on this.

In addition that the job template is loaded in the external runner before
the user is changed. otherwise the cluster_files_directory needs to be
readable by all users.

bernt-matthias commented:
@natefoo  @bgruening  still struggling with git. Since I was unable to rebase my branch to an updated dev I created a new PR with my changes. The old one (https://github.com/galaxyproject/galaxy/pull/4275) can be closed.

jmchilton commented on scripts/drmaa_external_runner.py:
```diff
@@ -116,15 +121,20 @@ def set_user(uid, assign_all_groups):
-    set_user(userid, assign_all_groups)
+    # load JSON job template data before changing the user
+    # then the pbs cluster_files_directory does not need to 
+    # be readable by all users
```
Yes - really nice!

bernt-matthias replied:
:)

jmchilton committed:
Reformat (PEP8) and touchup univa runner / PR changes.

jmchilton committed:
Merge remote-tracking branch 'jmchilton/dev' into topic/univa-drmaa

jmchilton commented:
Opened [a PR against your branch](https://github.com/bernt-matthias/galaxy/pull/1) to touch up the formatting and merge the latest dev to hopefully fix some of these tests. Can you merge it or give me permission to push these sorts of things to your branch directly?

Would you be willing to pull the checks of standard error for memory tool errors out of the runner? Like these lines:

```
+            # TODO 1 should be moved to runners/util and 2 should be implemented via the stdio tag of the tools 
 +            memerrors = set(["xrealloc: cannot allocate",
 +                 "MemoryError",
 +                 "std::bad_alloc",
 +                 "java.lang.OutOfMemoryError: Java heap space",
 +                 "Out of memory!"])
 +            toolmemerrors = set(["Out of memory" # data_manager_hisat2 
 +                 ])
 +            memviolation = util.grep_tail( ajs.error_file, memerrors | toolmemerrors, MEMORY_LIMIT_SCAN_SIZE )
```

I think the TODO there is correct and the tools should be annotating it - e.g. https://github.com/galaxyproject/galaxy/issues/3107. We could even setup up some common set of checks and provide and easy option for tool authors to quickly check for them but it isn't something I think we should be doing in one particular job runner. I also thing that there could be some tools that emit those things even for perfectly fine runs - e.g. a tool that tries one approach until it runs out of memory and if it does switches to another.

bernt-matthias commented:
@jmchilton I invited you as collaborator of my fork. 

We can move the stderr checks implemented in `memerrors` somewhere else. My idea is that its still a good idea to run them for every job. This could be done in the runners or more highlevel (which would be nice). The `toolmemerrors` could be moved to the tool's stdio checks, but currently the stdio checks of the tools seem not to allow resubmission of jobs (which is a really useful features which I use to avoid specifying memory and time requirements for every tool).

I haven't thought about jobs that are successful despite of such errors, I guess the simplest solution would be that within `<command>` some  post-processing  of stderr is done, ie. tool authors need to take care.

bernt-matthias commented:
Just had the idea that instead of the tool definition `<stdio>` way one could use some code in the `<command>` to set a flag indicating job resubmission (because of out of memory or out of time). 

For instance  data_manager_hisat2 could just echo `MemoryError` to stderr if it finds `"Out of memory"`, or use sed on stderr). 

We could just define a flag that wont appear in any tool, e.g. something like  `GALAXY_TOOL_OUT_OF_MEMORY`.

jmchilton commented:
@bernt-matthias I've opened a PR that allows tools to describe out of memory detection criteria here https://github.com/galaxyproject/galaxy/pull/5196 - want to give me your thoughts?

bernt-matthias committed:
bug fix parallel jobs available mem * slots

in parallel environments the requested vmem needs to be multiplied
by the number of granted slots. the number of slots is now parsed from
the qacct output (or qwait).

bernt-matthias committed:
added memory statement for UNIVA

- uses SGE_HGR_h_vmem
- replaces gigabytes (G$) by *1024
- multiplies with GALAXY_MEMORY_MB (i.e. needs to be computed before)
- makes it an integer

bernt-matthias committed:
removed multiplication with GALAXY_SLOTS

bernt-matthias commented:
Just added a memory statement for the UNIVA runner. Does somebody know if this is per thread or overall?

bernt-matthias committed:
restored GALAXY_MEMORY_MB as overall

jmchilton commented:
@bernt-matthias Thanks for sticking with this and sorry for the repeated delays, any chance you can merge with dev now that the memory stuff is merged to fix the conflict. There are also some linting issues with whitespace that you can see if you click on the Travis tests above. There is also a unit test failures stemming from reclaim_ownership interface changes between jobs and job wrappers - probably sticking a no-op ``reclaim_ownership`` method on this stub https://github.com/galaxyproject/galaxy/blob/dev/test/unit/jobs/test_runner_local.py#L121 should fix it.

That test file can be ran locally with:

```
. .venv/bin/activate
nosetests test/unit/jobs/test_runner_local.py
```

bernt-matthias committed:
used new variable in mem statement

bernt-matthias commented:
OK. Did a rebase. Is this what you requested? If there are any linting issues I will fix them as soon as travis finished.

I will see if I understand the failed test.

bernt-matthias commented:
Test seems to run now. Maybe related to the rebase..?

```
nosetests test/unit/jobs/test_runner_local.py
..........
----------------------------------------------------------------------
Ran 10 tests in 5.372s

OK

```

bernt-matthias committed:
fixed linter issues

bernt-matthias committed:
fixed linter issues

bernt-matthias committed:
fixed more linter issues

bernt-matthias committed:
fixed more linter issues

bernt-matthias committed:
more linting issues

bernt-matthias committed:
more linting issues

bernt-matthias committed:
linting fix

bernt-matthias committed:
linting fix

bernt-matthias commented:
All linting errors should be fixed now. Still unsure about the faling unit test. I guess it works locally because I did a rebase and no merge. Any comments?

jmchilton committed:
Remove comment with question - the answer is yes I believe.

jmchilton committed:
Remove comment with question - the answer is yes I believe.

jmchilton committed:
Fix failing unit test.

jmchilton commented:
I'm not sure what is up - but that unit test definitely fails for me the same way locally. The fix I pushed to your branch fixes it for me.

For the rest of the PR, I looks really good me. I was still hoping that #5196 would mean we could make:

```
elif memviolation or mem_wasted > mem_granted * slots:
```

just

```
elif mem_wasted > mem_granted * slots:
```

and the variables and computation making up ``memviolation``. Are you willing to do that or do you still think we should keep that logic here also for now?

bernt-matthias commented:
I'm open for moving or removing this .. preferring the former. 

Can you imagine a place where a check for such language specific memory errors could go such that also other/all schedulers profit? I still think that it would be a lot of redundancy if checks for language specific errors need to be implemented by every tool. 

On our cluster the checks seem to perform quite well giving me the opportunity to have a quite simple job_conf: we have 6 destinations that have resource requirements as combinations of 
- 1hour, 1day, and 1 week
- normal-memory, high memory
By detecting memory and runtime violations jobs are just escalated accordingly. I find this quite handy because only a few tools need special rules (usually that they enter not at the lowest requirements).

bernt-matthias committed:
added the description of known problem

bernt-matthias commented:
Just found a (small) problem and documented a workaround. The problem occurs when the jobs reach the queues soft limit. Workaround: configure destinations such that jobs never reach the soft limit.

jmchilton commented on lib/galaxy/jobs/runners/__init__.py:
```diff
@@ -389,6 +389,7 @@ def _find_container(
         )
 
     def _handle_runner_state(self, runner_state, job_state):
+        job_state.job_wrapper.reclaim_ownership()
```
Is this line actually needed? Maybe add a comment about why just reclaiming in finish is not enough?

bernt-matthias replied:
Actually I don't know if this is needed, and I can't remember that I did any changes to `lib/galaxy/jobs/runners/__init__.py` (but it has been a while ago). It could be as well caused by the problems with git I had here: https://github.com/galaxyproject/galaxy/pull/4275

The function seems to be called only from `fail_job()` and `_finish_or_resubmit_job()` [and 1x from ./galaxy/jobs/runners/kubernetes.py]. _finish_or_resubmit_job() is called by `finish_job()` which calls `reclaim_ownership` itself (so for this line it seems redundant). 

Concluding, I don't know if its necessary, but I know that it works... any comments?

bernt-matthias replied:
Did some more research: I did the change (https://github.com/bernt-matthias/galaxy/commit/f842bc5e6bde81d1fb4e02432ad43fb3f52c496e) probably in response to https://github.com/galaxyproject/galaxy/issues/4308.

bernt-matthias replied:
https://github.com/galaxyproject/galaxy/pull/6996

jmchilton commented:
Hmm... how about this. Can we agree that we should only check for error messages on failed jobs - because passing jobs might really pass despite those messages being in the standard error or output?

So if you wanted the logic to apply to all job runners - I might add it toward the end of ``check_output`` in lib/galaxy/jobs/output_checker.py. 

Some check for jobs already marked as errors:

```python
   if state == DETECTED_JOB_STATE.GENERIC_ERROR:
      oom_checking = getattr(tool.app.config, "enable_beta_aggressive_oom_checking")
      if oom_checking:
          # recheck the streams now...
          if oom:
              state = DETECTED_JOB_STATE.OUT_OF_MEMORY_ERROR

    # rest of this is unchanged...

    # Store the modified stdout and stderr in the job:
    if job is not None:
        job.set_streams(stdout, stderr)
```

bernt-matthias commented:
Hmm. About the memory checks. I can't imagine a scenario where such a message in the stderr would be OK (but I'm sure that it exists :smiley:). So I think that enabling it for all jobs won't hurt. On the other hand when a such a message occurs the application should be crashed already anyway (except if the programmer catches the exception and prints that it happened .. et voila there is the scenario that I was looking for :laughing:  BUT: the function checks only the tail of the stderr and finding an error there indicates that the error could not be handled). 

Concluding, 
- your suggestion (apply it to failing jobs only) seems the safe way to go (and improves the current state of the art)
- but I guess that applying it to all jobs should also work in almost all cases. 

So: I don't know ...

bernt-matthias commented:
Just found a small bug. There was a job that was apparently killed after a short time by the user. Then the runner was not able to find information on the job by qacct or qstat. Need to check with our admins how this is possible. 

Here in the relevant part of the log 

```
galaxy.jobs.runners.drmaa INFO 2018-04-26 11:57:09,874 (3781) queued as 4659220
galaxy.jobs DEBUG 2018-04-26 11:57:09,875 (3781) Persisting job destination (destination id: eve_short)
galaxy.jobs.runners.drmaa DEBUG 2018-04-26 11:57:10,128 (3781/4659220) state change: job is queued and active

galaxy.jobs.handler DEBUG 2018-04-26 11:57:18,143 Stopping job 3781:
galaxy.jobs.handler DEBUG 2018-04-26 11:57:18,143 stopping job 3781 in drmaa runner
galaxy.jobs.runners.drmaa DEBUG 2018-04-26 11:57:18,203 (3781/4659220) Removed from DRM queue at user's request
galaxy.jobs.runners.drmaa DEBUG 2018-04-26 11:57:19,473 (3781/4659220) state change: job finished normally
```
And then I get repeated error: 

```
galaxy.jobs.runners.univa ERROR 2018-04-26 12:02:29,765 `qacct -j 4659220` returned 1, stderr: error: job id 4659220 not found
Traceback (most recent call last):
  File "/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/jobs/runners/univa.py", line 484, in _get_drmaa_state
    wstate = self._get_drmaa_state_wait(job_id, ds, extinfo)
  File "/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/jobs/runners/univa.py", line 384, in _get_drmaa_state_wait
    raise DRMAAWaitUnusable()
DRMAAWaitUnusable
galaxy.jobs.runners ERROR 2018-04-26 12:02:29,787 Unhandled exception checking active jobs
Traceback (most recent call last):
  File "/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/jobs/runners/__init__.py", line 595, in monitor
    self.check_watched_items()
  File "/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/jobs/runners/drmaa.py", line 327, in check_watched_items
    if self._complete_terminal_job(ajs, drmaa_state=state) is not None:
  File "/gpfs1/data/galaxy_server/galaxy-dev/lib/galaxy/jobs/runners/univa.py", line 71, in _complete_terminal_job
    time_wasted = extinfo["time_wasted"]
KeyError: 'time_wasted'
```

Need to handle this. But first I ask out admins what exactly happened there.

bernt-matthias committed:
check for memory violations only for failed jobs

bernt-matthias committed:
univa runner: removed duplicated function

bernt-matthias committed:
drmaa univa: added try block around external run script

since unsuccessful job submission raises an error
this is needed, before this crashed the job queue

bernt-matthias committed:
Merge branch 'topic/univa-drmaa' into topic/univa-drmaa

bernt-matthias committed:
Merge pull request #1 from jmchilton/topic/univa-drmaa

Merge latest dev and touchup formatting.

bernt-matthias commented:
Hi, just though about the open issue for jobs that are deleted by the user before the scheduler had a chance to precess them. 

Is there somewhere a flag in the job data structure that indicates the job as deleted? Then I could accept the failing qstat/qacct.

natefoo commented:
It'll go through states `Job.states.DELETED_NEW` and `Job.states.DELETED` but there's no guarantee that it will end on `Job.states.DELETED`. You could look through the job's state history (`<job>.state_history`) but that might be a little tedious. If you wanted to add a `deleted` boolean column to `Job` that seems reasonable to me.

bernt-matthias committed:
univa: memchecks and unused functions removed

bernt-matthias commented:
Thanks @natefoo: I guess `ajs.job_wrapper.get_state() in [Job.states.DELETED, Job.states.DELETED_NEW]` might work. 

In order to finish this I need to update my galaxies in order to be able to test it (currently I can't stop jobs: https://github.com/galaxyproject/galaxy/pull/6912). I hope to find some time next week.

bernt-matthias commented:
@jmchilton I finally decided to remove all memviolation tests (in favour of per tool tests / aggressive mode)

bernt-matthias commented:
replaced https://github.com/galaxyproject/galaxy/pull/7004
